src/lightning_lite/plugins/collectives/torch_collective.py

```diff
@@ -0,0 +1,169 @@
+import datetime
+import os
+from typing import Any, List, Optional, Union
+
+import torch
+import torch.distributed as dist
+from typing_extensions import Self
+
+from lightning_lite.plugins.collectives.collective import Collective
+from lightning_lite.utilities.imports import _TORCH_GREATER_EQUAL_1_10
+from lightning_lite.utilities.types import CollectibleGroup, ReduceOp
+
+if dist.is_available():
+    from torch.distributed.constants import default_pg_timeout
+else:
+    default_pg_timeout = datetime.timedelta(seconds=1800)
+
+
+class TorchCollective(Collective):
+    def __init__(self) -> None:
+        if not dist.is_available():
+            raise RuntimeError("Torch distributed is not available.")
+        self._set_addr = False
+        self._set_port = False
+        super().__init__()
+
+    @property
+    def rank(self) -> int:
+        return dist.get_rank(self.group)
+
+    @property
+    def world_size(self) -> int:
+        return dist.get_world_size(self.group)
+
+    def broadcast(self, tensor: torch.Tensor, src: int) -> torch.Tensor:
+        dist.broadcast(tensor, src, group=self.group)
+        return tensor
+
+    def all_reduce(self, tensor: torch.Tensor, op: Union[str, ReduceOp] = "sum") -> torch.Tensor:
+        op = self._convert_to_native_op(op)
+        dist.all_reduce(tensor, op=op, group=self.group)
+        return tensor
+
+    def reduce(self, tensor: torch.Tensor, dst: int, op: Union[str, ReduceOp] = "sum") -> torch.Tensor:
+        op = self._convert_to_native_op(op)
+        dist.reduce(tensor, dst, op=op, group=self.group)
+        return tensor
+
+    def all_gather(self, tensor_list: List[torch.Tensor], tensor: torch.Tensor) -> List[torch.Tensor]:
+        dist.all_gather(tensor_list, tensor, group=self.group)
+        return tensor_list
+
+    def gather(
+        self, tensor: torch.Tensor, gather_list: Optional[List[torch.Tensor]] = None, dst: int = 0
```

Comment:
with an optional `gather_list` the API is not same. Can we do anything around that?
Same concern about returntype as above